presto-native-execution/presto_cpp/main/runtime-metrics/prometheus-reporter/PrometheusReporter.cpp

```diff
@@ -0,0 +1,229 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "presto_cpp/main/runtime-metrics/prometheus-reporter/PrometheusReporter.h"
+
+#include <prometheus/collectable.h>
+#include <prometheus/counter.h>
+#include <prometheus/gauge.h>
+#include <prometheus/histogram.h>
+#include <prometheus/registry.h>
+#include <prometheus/summary.h>
+#include <prometheus/text_serializer.h>
+
+namespace facebook::presto::prometheus {
+
+using namespace ::prometheus;
+
+static constexpr std::string_view kSummarySuffix("_summary");
+
+struct PrometheusReporter::Impl {
+  Impl(const Labels& labels) {
+    registry = std::make_shared<::prometheus::Registry>();
+    for (auto& itr : labels) {
+      this->labels[itr.first] = itr.second;
+    }
+  }
+
+ public:
+  ::prometheus::Labels labels;
+  std::shared_ptr<::prometheus::Registry> registry;
+};
+
+PrometheusReporter::PrometheusReporter(
+    const std::map<std::string, std::string>& labels) {
+  impl_ = std::make_shared<Impl>(labels);
+}
+void PrometheusReporter::registerMetricExportType(
+    const char* key,
+    facebook::velox::StatType statType) const {
+  if (registeredMetricsMap_.count(key)) {
+    // Already registered;
+    VLOG(1) << "Trying to register already registered metric " << key;
+    return;
+  }
+  // '.' is replaced with '_'.
+  std::string displayKey = std::string(key);
+  std::replace(displayKey.begin(), displayKey.end(), '.', '_');
+  switch (statType) {
+    case facebook::velox::StatType::COUNT: {
+      // A new MetricFamily object is built for every new metric key.
+      auto& counterFamily =
+          BuildCounter().Name(displayKey).Register(*impl_->registry);
+      auto& counter = counterFamily.Add(impl_->labels);
+      registeredMetricsMap_.emplace(
+          std::string(key), StatsInfo{statType, &counter});
+      break;
+    }
+    case facebook::velox::StatType::SUM:
+    case facebook::velox::StatType::AVG:
+    case facebook::velox::StatType::RATE: {
+      auto& gaugeFamily =
+          BuildGauge().Name(displayKey).Register(*impl_->registry);
+      auto& gauge = gaugeFamily.Add(impl_->labels);
+      registeredMetricsMap_.emplace(
+          std::string(key), StatsInfo{statType, &gauge});
+      break;
+    }
+    default:
+      VELOX_UNSUPPORTED(
+          "Unsupported metric type {}", static_cast<int>(statType));
+  }
+}
+
+void PrometheusReporter::registerMetricExportType(
+    folly::StringPiece key,
+    facebook::velox::StatType statType) const {
+  registerMetricExportType(key.toString().c_str(), statType);
+}
+
+void PrometheusReporter::registerHistogramMetricExportType(
+    const char* key,
+    int64_t bucketWidth,
+    int64_t min,
+    int64_t max,
+    const std::vector<int32_t>& pcts) const {
+  if (registeredMetricsMap_.count(key)) {
```

Comment:
Lets log once we register metrics, 

LOG(INFO) << "Registered Histogram for key: " << keyStr;

same for others